### ISSUE TITLE
cont_cat_split does not support pandas types / fastbook 09_tabular.ipynb fails

### ISSUE THREAD

**Issue by chrismilleruk** [fastai/fastai#3156]
**Prerequisites**
- [x] Search [the forums](https://forums.fast.ai) for the error message you received. Ask on the forums to see if someone else has seen a similar issue already and knows how to solve it. 
  - https://forums.fast.ai/t/problem-with-cont-cat-split-in-google-collab/84402/3
  - https://github.com/fastai/fastbook/issues/382
- [x] Please confirm you have the latest versions of fastai, fastcore, and nbdev prior to reporting a bug.

**Describe the bug**
`cont_cat_split()` fails in `fastai/fastbook/blob/master/clean/09_tabular.ipynb`:
`TypeError: Cannot interpret ‘UInt32Dtype()’ as a data type`

**To Reproduce**
Requires pandas >= 1.1.0 and fastai >= 2.2.0

1. Open a clean version of fastai/fastbook on Colab or Gradient
2. Open 09_tabular.ipynb (either clean or regular)
3. Run all cells.

Simplified Repro:
```python
df = pd.DataFrame({ 'd1_date': ['2021-02-09'] })
df = add_datepart(df, 'd1_date')
cont_names, cat_names = cont_cat_split(df)
```

**Expected behavior**
Notebook should complete without errors.

**Error with full stack trace**
```
TypeError Traceback (most recent call last)
in ()
----> 1 cont,cat = cont_cat_split(df, 1, dep_var=dep_var)

1 frames
/usr/local/lib/python3.6/dist-packages/numpy/core/numerictypes.py in issubdtype(arg1, arg2)
386 “”"
387 if not issubclass_(arg1, generic):
–> 388 arg1 = dtype(arg1).type
389 if not issubclass_(arg2, generic):
390 arg2 = dtype(arg2).type

TypeError: Cannot interpret ‘UInt32Dtype()’ as a data type
```

**Additional context**
`add_datepart` generates fields with numpy types but `week` can be taken from `isocalendar()` and is returning as `pd.UInt32Dtype` instead of `int64`/`float64`
`cont_cat_split` uses `np.issubdtype` which does not play well with `pandas` types (known issue: https://github.com/numpy/numpy/issues/9474)

Related issues and PRs:
https://github.com/fastai/fastbook/issues/382
https://github.com/fastai/fastai/issues/3114
https://github.com/fastai/fastai/pull/3116
https://github.com/fastai/fastai/pull/3117
https://github.com/fastai/fastai/issues/2583

**Comment by Isaac-Flath** [comment]
I am unable to replicate the simplified repro.  No errors for me

![contcatissue](https://user-images.githubusercontent.com/6256508/104676945-0b5eea00-56ae-11eb-8773-8571d023e8d3.png)

**Comment by chrismilleruk** [comment]
Thanks @Isaac-Flath - As of this morning, I'm still able to repro on both gradient and colab.

Can you print your versions and dtypes please?

<img width="416" alt="Screenshot 2021-01-15 at 08 39 47" src="https://user-images.githubusercontent.com/637130/104701973-45aba400-570d-11eb-9c32-895f3eac4012.png">

<img width="557" alt="Screenshot 2021-01-15 at 08 33 48" src="https://user-images.githubusercontent.com/637130/104700979-e5b4fd80-570c-11eb-870a-e82362bfaa90.png">

**Comment by Isaac-Flath** [comment]
Ah, it works in pandas version 1.0.5 but not the next version up (1.1.0) and I was on an old version of pandas (sorry!).

There is a comment in add_datepart (    # Pandas removed `dt.week` in v1.1.10).  The fix in add_datepart works but makes the week field a dtype not compatible with cont_cat_split (particularly the np.issubdtype line).

The easiest fix is to cast that field as an int.  I will put in a pr for this issue this evening.  In the meantime, you can modify the function and it should work.  This is the only line that changes:
    `week = int(field.dt.isocalendar().week) if hasattr(field.dt, 'isocalendar') else int(field.dt.week)`


### Fixed function
```python
def add_datepart(df, field_name, prefix=None, drop=True, time=False):
    "Helper function that adds columns relevant to a date in the column `field_name` of `df`."
    make_date(df, field_name)
    field = df[field_name]
    prefix = ifnone(prefix, re.sub('[Dd]ate$', '', field_name))
    attr = ['Year', 'Month', 'Day', 'Dayofweek', 'Dayofyear', 'Is_month_end', 'Is_month_start',
            'Is_quarter_end', 'Is_quarter_start', 'Is_year_end', 'Is_year_start']
    if time: attr = attr + ['Hour', 'Minute', 'Second']
    for n in attr: df[prefix + n] = getattr(field.dt, n.lower())
    # Pandas removed `dt.week` in v1.1.10
    week = int(field.dt.isocalendar().week) if hasattr(field.dt, 'isocalendar') else int(field.dt.week)
    df.insert(3, prefix+'Week', week)
    mask = ~field.isna()
    df[prefix + 'Elapsed'] = np.where(mask,field.values.astype(np.int64) // 10 ** 9,None)
    if drop: df.drop(field_name, axis=1, inplace=True)
    return df
```

**Comment by chrismilleruk** [comment]
Ah great, thanks for confirming, that's helpful.

I found that line and made a similar fix (I used `.astype(` but not sure if better). I also noticed that the other types were dependent on if all date values are present (`int64` as in the above example) vs if the column contained None (becomes `float64`)

This didn't solve the issue for Category and other Pandas types though:
```python
# Example with pandas types and generated columns
df = pd.DataFrame({'cat1': pd.Series(['l','xs','xl','s'], dtype='category'),
                    'ui32': pd.Series([1, 2, 3, 4], dtype='UInt32'),
                    'i64': pd.Series([1, 2, 3, 4], dtype='Int64'),
                    'f16': pd.Series([1, 2, 3, 4], dtype='Float64'),
                    'd1_date': ['2021-02-09', None, '2020-05-12', '2020-08-14'],
                    })
df = add_datepart(df, 'd1_date', drop=False)
df['cat1'].cat.set_categories(['xl','l','m','s','xs'], ordered=True, inplace=True)
cont_names, cat_names = cont_cat_split(df, max_card=0)
```
I found that changing `cont_cat_split` to use `pd.api.types.is_integer_dtype` over `np.issubtype` seems to be more reliable.

I created a PR (https://github.com/fastai/fastai/pull/3157) incorporating the two fixes above and would welcome your thoughts. 

Thanks

**Comment by Isaac-Flath** [comment]
Looks good to me!

**Comment by chrismilleruk** [comment]
Fix merged in, should land in 2.2.4
